rust/agama-dbus-server/src/questions.rs

```diff
@@ -1,162 +1,109 @@
     pub fn default_option(&self) -> &str {
-        match self.default_option {
-            Some(ref option) => option.as_str(),
-            None => "",
-        }
+        self.0.default_option.as_str()
     }
 
     #[dbus_interface(property)]
     pub fn answer(&self) -> &str {
-        &self.answer
+        &self.0.answer
     }
 
     #[dbus_interface(property)]
     pub fn set_answer(&mut self, value: &str) -> Result<(), zbus::fdo::Error> {
         // TODO verify if answer exists in options or if it is valid in other way
-        self.answer = value.to_string();
+        self.0.answer = value.to_string();
 
         Ok(())
     }
 }
 
-/// Specialized question for Luks partition activation
-#[derive(Clone, Debug)]
-pub struct LuksQuestion {
-    /// Luks password. Empty means no password set.
-    password: String,
-    /// number of previous attempts to decrypt partition
-    attempt: u8,
-    /// rest of question data that is same as for other questions
-    generic_question: GenericQuestion,
-}
+/// Mixin interface for questions that are base + contain question for password
+struct WithPasswordObject(questions::WithPassword);
 
-impl LuksQuestion {
-    fn device_info(device: &str, label: &str, size: &str) -> String {
-        let mut result = device.to_string();
-        if !label.is_empty() {
-            result = format!("{} {}", result, label);
-        }
-
-        if !size.is_empty() {
-            result = format!("{} ({})", result, size);
-        }
-
-        result
-    }
-
-    pub fn new(id: u32, device: String, label: String, size: String, attempt: u8) -> Self {
-        let msg = format!(
-            "The device {} is encrypted.",
-            Self::device_info(device.as_str(), label.as_str(), size.as_str())
-        );
-        let base = GenericQuestion::new(
-            id,
-            msg,
-            vec!["skip".to_string(), "decrypt".to_string()],
-            Some("skip".to_string()),
-        );
-
-        Self {
-            password: "".to_string(),
-            attempt,
-            generic_question: base,
-        }
-    }
-
-    pub fn base(&self) -> &GenericQuestion {
-        &self.generic_question
-    }
-}
-
-#[dbus_interface(name = "org.opensuse.Agama.Questions1.LuksActivation")]
-impl LuksQuestion {
+#[dbus_interface(name = "org.opensuse.Agama.Questions1.WithPassword")]
+impl WithPasswordObject {
     #[dbus_interface(property)]
     pub fn password(&self) -> &str {
-        self.password.as_str()
+        self.0.password.as_str()
     }
 
     #[dbus_interface(property)]
     pub fn set_password(&mut self, value: &str) {
-        self.password = value.to_string();
-    }
-
-    #[dbus_interface(property)]
-    pub fn attempt(&self) -> u8 {
-        self.attempt
+        self.0.password = value.to_string();
     }
 }
 
 /// Question types used to be able to properly remove object from dbus
 enum QuestionType {
-    Generic,
-    Luks,
+    Base,
+    BaseWithPassword,
+}
+
+trait AnswerStrategy {
```

Comment:
I would expect a description of what this trait does and how is expected to be used. It can be in a different PR, as we agreed offline, but please do not forget about it.

Reply:
Moreover, perhaps it deserves its own `strategies` module.

Reply:
yeap, own module can happen easily with adding FileStrategy to get it from file. I just do not want to over-complicate it now

Reply:
OK, no problem. We can move later.